Add Integration test

CaioCordeiro opened the pull request:
This patch adds a integration test.
Changes : 
- Adds test folder and the integration test file;
- Adds the producer part to the integration test;
- Adds the consumer part of the integration test.
-  Add the integration test compilation script to the automake
so it can be runned with the 'make check' command;
- Add the test commands to the README.

CaioCordeiro committed:
Add integration test producer

This patch adds the producer integration test.
Changes:
- Adds test folder and the integration test file;
- Adds the producer part to the integration test.

CaioCordeiro committed:
Add integration test consumer

This patch adds the consumer part of the integration test.
The consumer reads the posted message from the producer and verify
its integrity.

CaioCordeiro committed:
Add integration test to automake

This patch adds the integration test to the automake.
Changes:
-  Add the integration test compilation script to the automake
so it can be runned with the 'make check' command;
- Add the test commands to the README.

CaioCordeiro committed:
Add parser unit test

This patch adds unit tests to all parser.h functions.
Functions tested:
- parser_config_create_object;
- parser_update_to_list;
- parser_data_create_object;
- parser_config_to_list;
- parser_queue_from_json_array;
- parser_request_to_list;
- parser_sensorid_to_json;
- parser_device_json_create;
- parser_auth_json_create;
- parser_unregister_json_create;
- parser_get_key_str_from_json_str;
- parser_is_key_str_or_null;

CaioCordeiro committed:
Add unit test to automake

This patch adds the unit test compilation script together with the
expected output to the "Makefile.am" so it can be runned with the
'make check' command.

CaioCordeiro committed:
Update README with the new test command

This patch adds the test documentation to the README.

alla3 commented on test/integration/integration_test.c:
```diff
@@ -0,0 +1,169 @@
+	{
+	case SIGINT:
+	case SIGTERM:
+
```
Remove the blank line

luigiluz commented on test/integration/integration_test.c:
```diff
@@ -0,0 +1,169 @@
+#include <knot/knot_types.h>
+#include <knot/knot_cloud.h>
+
+#include <assert.h>
```
I believe that this header is being included twice

CaioCordeiro replied:
Done

luigiluz commented on test/integration/integration_test.c:
```diff
@@ -0,0 +1,169 @@
+	}
+}
+
+static bool on_recived_msg(const struct knot_cloud_msg *msg,
```
```suggestion
static bool on_received_msg(const struct knot_cloud_msg *msg,
```

CaioCordeiro replied:
Done

luigiluz commented on test/integration/integration_test.c:
```diff
@@ -0,0 +1,169 @@
+	strcpy(token, "41aa0229342ecf8eb686cde53f739c1c3da9c1c5");
+	strcpy(id, "5b620bcd419afed7");
+
+
```
I don't think these two blank lines are necessary, if they aren't, please remove them.

CaioCordeiro replied:
Done

luigiluz commented on test/integration/integration_test.c:
```diff
@@ -0,0 +1,169 @@
+
+	err = knot_cloud_update_schema("DEVICE_ID", schema_queue);
+
+	l_queue_destroy(schema_queue,NULL);
```
```suggestion
	l_queue_destroy(schema_queue, NULL);
```

CaioCordeiro replied:
Done

luigiluz commented on test/integration/integration_test.c:
```diff
@@ -0,0 +1,169 @@
+	int err;
+
+	//READ START
+	knot_cloud_read_start("5b620bcd419afed7", on_recived_msg, NULL);
```
You're using this id multiple times in the entire code, maybe it would be nice to create a few defines to avoid using the same value a lot of times. (Maybe it would be nice to have all these constants as defines, please consider this if it won't cause you much trouble)

CaioCordeiro replied:
Done

luigiluz reviewed (changes requested):
Awesome PR! Just made a few remarks, let me know when you take a look at them!

luigiluz commented:
Also in the commit [bdbeb15](https://github.com/CESARBR/knot-cloud-sdk-c/pull/13/commits/bdbeb15149aa14b5209f9774eb4b5f3c81bf4aa7) there's a typo in the commit message body.

Please change "parto" to "part"

CaioCordeiro committed:
Fix create_device_item_cb parameter name

This patch changes the "struct l_queue *schema" parameter to
"struct l_queue *config" to match the code terminology.

agms-cesar commented on tests/scripts/integration_test.sh:
```diff
@@ -0,0 +1,68 @@
+DOCKER_NAME="rabbitmq-test-env"
+
+start_docker_env() {
+	CURR_TRY=0
```
you can name those local variables lower_case

agms-cesar commented on tests/scripts/integration_test.sh:
```diff
@@ -0,0 +1,68 @@
+	RABBITMQ_URL="localhost:15672"
+	DOCKER_IMAGE="rabbitmq:3-management-alpine"
+
+	docker run -d --hostname "$DOCKER_HOSTNAME" --name "$DOCKER_NAME" \
```
you can capture the output of this command and validate if it ran successfully
